pkg/queue/concurrency_state.go

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"go.uber.org/zap"
+)
+
+// ConcurrencyStateHandler tracks the in flight requests for the pod. When the requests
+// drop to zero, it runs the `pause` function, and when requests scale up from zero, it
+// runs the `resume` function. If either of `pause` or `resume` are not passed, it runs
+// the respective local function(s). The local functions are the expected behavior; the
+// function parameters are enabled primarily for testing purposes.
+func ConcurrencyStateHandler(h http.Handler, logger *zap.SugaredLogger, pause, resume func(string) error, endpoint string) http.HandlerFunc {
+	logger.Info("Concurrency state endpoint set, tracking request counts, using endpoint ", endpoint)
+
+	type req struct {
+		w http.ResponseWriter
+		r *http.Request
+
+		done chan struct{}
+	}
+
+	reqCh := make(chan req)
+	doneCh := make(chan struct{})
+	go func() {
+		inFlight := 0
+
+		// This loop is entirely synchronous, so there's no cleverness needed in
+		// ensuring open and close dont run at the same time etc. Only the
+		// delegated ServeHTTP is done in a goroutine.
+		for {
+			select {
+			case <-doneCh:
+				inFlight--
+				if inFlight == 0 {
+					logger.Info("Requests dropped to zero")
+					err := pause(endpoint)
+					if err != nil {
+						logger.Errorf("Issue handling pause request: %s", err)
+					} else {
+						logger.Info("Request successfully processed")
+					}
+				}
+
+			case r := <-reqCh:
+				inFlight++
+				if inFlight == 1 {
+					logger.Info("Requests increased from zero")
+					err := resume(endpoint)
+					if err != nil {
+						logger.Errorf("Issue handling resume request: %s", err)
```

Comment:
ditto here, if we fail to resume and then send a request to the paused pod anyway that's not going to end well obviously. A panic is better than that, but we should be clear if we want to try retrying instead, or maybe the pod should fail a readiness/liveness check or something so k8s will replace it? (Some of this might well be follow-on PRs, just thinking out loud and making sure we don't lose this thread because I think it's important - and just logging definitely seems wrong I think).